res/axios_throttler.js

```diff
@@ -0,0 +1,109 @@
+(function(){
+    let oldPost = axios.post;
+
+    class AxiosThrottler {
+        constructor() {
+            this.debug = false;
+            this.requestCache = new Map();
+            // Minimum interval between "throttled" requests
+            this.throttleMs = 10 * 1000;
+            // Interval between retries for retriable requests
+            this.retryBackoffMs = 2 * 1000;
+            // Maxmimum number of *additional* retries for retriable requests
+            this.maxRetries = 2;
+        }
+
+        // Only throttle specific request types so that we don't accidentally break the
+        // game state by returning incorrect information.
+        isCacheableRequest(req) {
+            return req.url === "api/game/fetch_game_activity_date.php";
+        }
+
+        // Only retry specific request types so that we don't accidentally retry
+        // destructive operations.
+        isRetriableRequest(req) {
+            return req.url === "api/game/load_replay.php";
+        }
+
+        // Fulfill the request with cache-based "throttling" of duplicate requests within
+        // the throttle interval. This can return data that is up to throttleMs stale.
+        // Behavior is to:
+        // 1. always allow requests if they:
+        //     a. haven't been seen before (aren't in the cache)
+        //     b. have a cache entry that is stale (older than throttleMs)
+        // 2. return the cached value if the request was last made within throttleMs
+        // 3. coalesce concurrent requests onto a single outgoing request
+        doWithCacheThrottling(req) {
+            let key = JSON.stringify({url: req.url, data: req.data});
+            let cacheEntry = this.requestCache.get(key);
+
+            if (cacheEntry === undefined) {
+                if (this.debug) console.log("No cached response for throttlable request, proceeding.");
+            } else if (this.throttleMs < (Date.now() - cacheEntry.timeMs)) {
+                if (this.debug) console.log("Cached response for throttlable request is stale, proceeding.");
+            } else {
+                if (this.debug) console.log("Cached response is fresh enough, throttling.");
+                return cacheEntry.promise;
+            }
+
+            // Store the promise in the cache, not the value, so that concurrent requests
+            // will all block on a single outgoing request, rather than concurrently making requests
+            // and fighting over who can write to the cache.
+            let promise = oldPost(req.url, req.data, req.config);
+            this.requestCache.set(key, {promise: promise, timeMs: Date.now()});
+            return promise.then((res) => {
+                if (this.debug) console.log("Newly cached promise for:", key, "got value:", res);
+                return res;
+            });
+        }
+
+        // Fulfill the request with a limited backoff and retry loop on 503 errors.
+        doWithRetriesInternal(req, retriesRemaining) {
```

Comment:
A convention for internal functions is to name them with a preceding "_". 
i.e. "_doWithRetriesInternal()"